src/graph/ChcGraph.cc

```diff
@@ -599,4 +599,194 @@ std::unique_ptr<ChcDirectedHyperGraph> ChcDirectedHyperGraph::makeEmpty(Logic &
         std::move(predicateRepresentation),
         logic
         );
+}
+
+PTRef ChcDirectedGraph::translateEdge(DirectedEdge const & edge, LocationVarMap const & locationVarMap,
+                                      PositionVarMap const & positionVarMap) const {
+    TermUtils::substitutions_map substitutionsMap;
+    auto source = edge.from;
+    auto target = edge.to;
+    TimeMachine timeMachine(logic);
+
+    auto edgeVariables = getVariablesFromEdge(edge.id);
+
+    // TODO: prepare the substitution map in advance!
+    auto const & stateVars = edgeVariables.stateVars;
+    for (unsigned int i = 0; i < stateVars.size(); ++i) {
+        VarPosition varPosition{source, i};
+        auto it = positionVarMap.find(varPosition);
+        assert(it != positionVarMap.end());
+        substitutionsMap.insert({stateVars[i], it->second});
+    }
+
+    auto const & nextStateVars = edgeVariables.nextStateVars;
+    for (unsigned int i = 0; i < nextStateVars.size(); ++i) {
+        VarPosition varPosition{target, i};
+        auto it = positionVarMap.find(varPosition);
+        assert(it != positionVarMap.end());
+        substitutionsMap.insert({nextStateVars[i], timeMachine.sendVarThroughTime(it->second, 1)});
+    }
+
+    // Translate the constraint
+    PTRef translatedConstraint = TermUtils(logic).varSubstitute(edge.fla.fla, substitutionsMap);
+    PTRef sourceLocationVar = source == getEntry() ? logic.getTerm_true() : locationVarMap.at(source);
+    PTRef targetLocationVar = locationVarMap.at(target);
+    PTRef updatedLocation = [&]() {
+        vec<PTRef> args;
+        args.capacity(locationVarMap.size() * 2);
+        for (auto && entry : locationVarMap) {
+            if (entry.second != targetLocationVar) {
+                args.push(logic.mkNot(timeMachine.sendVarThroughTime(entry.second, 1)));
+            }
+            if (entry.second != sourceLocationVar) { args.push(logic.mkNot(entry.second)); }
+        }
+        return logic.mkAnd(std::move(args));
+    }();
+    // We used to add equalities that restricted the values of all variables other than target ones to their default
+    // values. The paper uses frame equalities that keep the values from previous step. Now, we do not restrict
+    // the values of these variables in any way.
+    // This is sound because we still force the right variables to be considered using the location variables.
+    vec<PTRef> components{sourceLocationVar, translatedConstraint, timeMachine.sendVarThroughTime(targetLocationVar, 1),
+                          updatedLocation};
+    return logic.mkAnd(std::move(components));
+}
+
+WitnessInfo ChcDirectedGraph::contractConnectedVertexes(std::vector<EId> edges) {
+    assert(edges.size() >= 2);
+    assert(getSource(edges[0]) == getTarget(edges[edges.size() - 1]));
+    TimeMachine timeMachine(logic);
```

Comment:
Mb it's better to create `timeMachine` and `termUtils` once per graph during construction then call constructor again and again?

There are 23 constructor calls to termUtils just in this file!

Reply:
`TimeMachine` and `TermUtils` are trivial objects, they just store a reference to `Logic`.
It is OK to create them on demand.